url3f8d6440dd_huiwq1990_CognativeModel_tgz-pJ8-clarion_samples_HelloWorldJ8/src/clarion/system/ACS.java

```diff
@@ -0,0 +1,1521 @@
+							if(rs > vIRL)
+								vIRL = rs;
+						}
+						else
+						{
+							double ru = r.getUtility();
+							if(ru > vIRL)
+								vIRL = ru;
+						}
+					}
+				}
+			}
+			//Fixed Rules
+			double vFR = 0;
+			rc = FRRuleStore.getRules(a);
+			if(rc != null)
+			{
+				for(AbstractRule r : rc)
+				{
+					r.setCurrentInput(CurrentInput);
+					if(r.checkEligibility(CurrentTimeStamp))
+					{
+						usageCandidates.add(r);
+						if(r.SELECTION_TYPE == AbstractRule.SelectionTypes.SUPPORT)
+						{
+							double rs = r.getSupport(CurrentInput);
+							if(rs > vFR)
+								vFR = rs;
+						}
+						else
+						{
+							double ru = r.getUtility();
+							if(ru > vFR)
+								vFR = ru;
+						}
+					}
+				}
+			}
+			//Combine
+			a.setTLSelectionMeasure((wRER * vRER) + (wIRL * vIRL) + (wFR * vFR));
+			a.setFinalSelectionMeasure(a.getTLSelectionMeasure() + (wBL * a.getBLSelectionMeasure()));
+		}
+	}
+	
+	/**
+	 * Performs bottom level learning, rule refinement, and match statistic updates based on the 
+	 * specified new input. If this method is used, learning is performed without any feedback.
+	 * @param NewInput The new input that was perceived.
+	 * @param TimeStamp The current time stamp.
+	 */
+	protected void learn (DimensionValueCollection NewInput, long TimeStamp)
+	{
+		if(EM != null)
+			ChosenAction = EM.getPerformedAction(CurrentTimeStamp);
+		updateInputSpace(NewInput.values());
+		if(ChosenAction != null)
+		{
+			updateMatchAllStats(ChosenAction, ChosenAction.getFinalSelectionMeasure(), MatchCalculator);
+			updateRules(ChosenAction.getFinalSelectionMeasure(), TimeStamp);
+			updateBottomLevel(NewInput, ChosenAction.getFinalSelectionMeasure());
+		}
+	}
+	
+	/**
+	 * Performs bottom level learning, rule refinement, and match statistic updates based on the 
+	 * specified new input.
+	 * @param NewInput The new input that was perceived.
+	 * @param feedback The feedback received.
+	 * @param TimeStamp The current time stamp.
+	 */
+	protected void learn (DimensionValueCollection NewInput, double feedback, long TimeStamp)
+	{
+		if(EM != null)
+			ChosenAction = EM.getPerformedAction(CurrentTimeStamp);
+		updateInputSpace(NewInput.values());
+		if(ChosenAction != null)
+		{
+			updateMatchAllStats(ChosenAction, feedback, MatchCalculator);
+			updateRules(feedback, TimeStamp);
+			updateBottomLevel(NewInput, feedback);
+		}
+	}
+	
+	/**
+	 * Updates the bottom level implicit representations (if they support online learning) given the 
+	 * new sensory information and feedback (if any was received). This method also updates the
+	 * match statistics for the bottom level components and the bottom level module itself.
+	 * <p>
+	 * If any external rules can be extracted from the bottom level, this method extracts them.
+	 * @param NewInput The new input.
+	 * @param feedback The feedback.
+	 */
+	private void updateBottomLevel (DimensionValueCollection NewInput, double feedback)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 25. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 666-732 into a separate method.